File: hackerrank.com/minimum-swaps-2.py
#!/bin/python3
#https://www.hackerrank.com/challenges/minimum-swaps-2

import math
import os
import random
import re
import sys

# Complete the minimumSwaps function below.
def minimumSwaps_polynomial(arr):
    count = 0
    i = 0

    for i in range(len(arr)):
        if arr[i] != i + 1:
            for j in range(i + 1, len(arr)):
                if (arr[j] == i + 1):
                    arr[i], arr[j] = arr[j], arr[i]
                    count += 1

    return count

def minimumSwaps(arr):
    count = 0
    i = 0

    while i < len(arr):
        if arr[i] == i + 1:
            i += 1
            continue
        
        t = arr[arr[i] - 1]
        arr[arr[i] - 1] = arr[i]
        arr[i] = t
        count += 1

    return count


if __name__ == '__main__':
    #fptr = open(os.environ['OUTPUT_PATH'], 'w')
    #n = int(input())
    arr = [1, 3, 2, 4] #list(map(int, input().rstrip().split()))
    res = minimumSwaps(arr)
    #fptr.write(str(res) + '\n')
    #fptr.close()

File: hackerrank.com/minimum-swaps-2.go
package main

import (
    "bufio"
    "fmt"
    "io"
    "os"
    "strconv"
    "strings"
)

// Complete the minimumSwaps function below.
func minimumSwaps(arr []int32) int32 {
    var count = int32(0)

    for i := int32(0); i < int32(len(arr)); i++ {
        if arr[i] != i + 1 {
            for j := i + 1; j < int32(len(arr)); j++ {
                if arr[j] == i + 1 {
                    var temp = arr[j]
                    arr[j] = arr[i]
                    arr[i] = temp

                    count++
                }
            }
        }
	}
	
	return count
}

func main() {
    reader := bufio.NewReaderSize(os.Stdin, 1024 * 1024)

    stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
    checkError(err)

    defer stdout.Close()

    writer := bufio.NewWriterSize(stdout, 1024 * 1024)

    nTemp, err := strconv.ParseInt(readLine(reader), 10, 64)
    checkError(err)
    n := int32(nTemp)

    arrTemp := strings.Split(readLine(reader), " ")

    var arr []int32

    for i := 0; i < int(n); i++ {
        arrItemTemp, err := strconv.ParseInt(arrTemp[i], 10, 64)
        checkError(err)
        arrItem := int32(arrItemTemp)
        arr = append(arr, arrItem)
    }

    res := minimumSwaps(arr)

    fmt.Fprintf(writer, "%d\n", res)

    writer.Flush()
}

func readLine(reader *bufio.Reader) string {
    str, _, err := reader.ReadLine()
    if err == io.EOF {
        return ""
    }

    return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
    if err != nil {
        panic(err)
    }
}
